ckanext/data_qld/validation.py

```diff
@@ -21,16 +24,176 @@ def scheming_choices(field, schema):
+        OneOf = tk.get_validator('OneOf')
+
         return OneOf([c['value'] for c in field['choices']])
 
     def validator(value):
-        if value is missing or not value:
+        if value is tk.missing or not value:
             return value
         choices = sh.scheming_field_choices(field)
         for c in choices:
             #  We want the value check to be case insensitive
             if value.upper() == c['value'].upper():
                 return value
-        raise Invalid(_('unexpected choice "%s"') % value)
+        raise tk.Invalid(tk._('unexpected choice "%s"') % value)
 
     return validator
+
+
+def process_schema_fields(key, data, errors, context):
+    schema_from_upload_request = read_schema_from_request()
+
+    if schema_from_upload_request:
+        data[key] = schema_from_upload_request
+        _clear_pseudo_fields(data)
+        return
+
+    schema_from_upload_file = read_schema_from_file(data)
+    if schema_from_upload_file:
+        data[key] = schema_from_upload_file
+        _clear_pseudo_fields(data)
+        return
+
+    schema_from_url = get_schema_from_url(key, data, errors)
+    if schema_from_url:
+        data[key] = schema_from_url
+        _clear_pseudo_fields(data)
+        return
+
+    schema_from_json = get_schema_from_json(data)
+    if schema_from_json:
+        data[key] = schema_from_json
+        _clear_pseudo_fields(data)
+        return
+
+
+def read_schema_from_request():
+    try:
+        tk.request.files
+    except TypeError:
+        # working outside context, cli or tests
+        return
+
+    if tk.request.files.get("schema_upload"):
+        schema_upload = tk.request.files.get("schema_upload")
+        return _get_underlying_file(schema_upload).read()
+
+
+def read_schema_from_file(data):
+    schema_upload_key = ("schema_upload", )
+    schema_upload = data.get(schema_upload_key)
+
+    if isinstance(schema_upload, ALLOWED_UPLOAD_TYPES) \
+        and schema_upload and schema_upload.filename:
+        data[schema_upload_key] = ""
+        return _get_underlying_file(schema_upload).read()
+
+
+def get_schema_from_url(key, data, errors):
+    schema_url_key = ("schema_url", )
+    value = data.get(schema_url_key)
+
+    if not value:
+        return
+
+    if (value and not isinstance(value, string_types)
+            or not value.lower()[:4] == u'http'):
+        data[schema_url_key] = ""
+        err_msg = tk._('Must be a valid URL "{}"').format(value)
+        raise tk.Invalid(err_msg)
+
+    return value
+
+
+def get_schema_from_json(data):
+    schema_json_key = ("schema_json", )
+    value = data.get(schema_json_key)
+
+    if value:
+        data[schema_json_key] = ""
+        return value
+
+
+def _clear_pseudo_fields(data):
+    schema_json_key = ("schema_json", )
+    schema_url_key = ("schema_url", )
+    schema_upload_key = ("schema_upload", )
+
+    data[schema_json_key] = data[schema_url_key] = data[schema_upload_key] = ""
+
+
+ALIGNED = True
```

Comment:
Constants should go at the top of the module, before any functions.

Reply:
Will be moved in constants. Thanks.

Reply:
Will be fixed and pushed by Alex shortly.